app/src/main/java/ch/epfl/sdp/appart/AdActivity.java

```diff
@@ -192,18 +211,125 @@ public boolean onCreateOptionsMenu(Menu menu) {
         return true;
     }
 
+    /**
+     * Adds ad to favorites, shows a toast with the action result
+     */
     @Override
     public boolean onOptionsItemSelected(MenuItem item) {
-        // TODO check the user is logged in
         if (item.getItemId() == R.id.action_add_favorite) {
-            database.getUser(login.getCurrentUser().getUserId()).thenAccept(u -> {
-                u.addFavorite(adId);
-                database.updateUser(u);
+            CompletableFuture<Void> addRes = addNewFavorite();
+            addRes.exceptionally(e -> {
+                Toast.makeText(this, e.getMessage(),
+                        Toast.LENGTH_SHORT).show();
+                return null;
             });
+            addRes.thenAccept(res ->
+                    Toast.makeText(this, R.string.favSuccess_Ad,
+                            Toast.LENGTH_SHORT).show()
+            );
             return true;
         }
         return super.onOptionsItemSelected(item);
     }
 
+    /**
+     * Check if user is logged in and add new favorite ad to user
+     */
+    private CompletableFuture<Void> addNewFavorite() {
+        CompletableFuture<Void> result = new CompletableFuture<>();
+        // check that the user is logged in
+        User user = login.getCurrentUser();
+        if (user == null) {
+            result.completeExceptionally(
+                    new UnsupportedOperationException(getString(R.string.userNotLoggedIn_Ad))
+            );
+            return result;
+        }
+        CompletableFuture<User> userRes = database.getUser(user.getUserId());
+        userRes.exceptionally(e -> {
+            result.completeExceptionally(
+                    new DatabaseServiceException(getString(R.string.favFail_Ad)));
+            return null;
+        });
+        userRes.thenAccept(u -> saveFavorite(result, u));
+        return result;
+    }
 
+    /**
+     * Save new ad id in database and if successful save ad locally
+     */
+    private void saveFavorite(CompletableFuture<Void> result, User user) {
+        user.addFavorite(adId);
+        CompletableFuture<Boolean> updateRes = database.updateUser(user);
+        updateRes.exceptionally(e -> {
+            result.completeExceptionally(
+                    new DatabaseServiceException(
+                            getString(R.string.favFail_Ad)));
+            return null;
```

Comment:
Just as curiosity, why don't we return the content of e that could provide us with maybe more information about the failure ?

Reply:
I wasn't sure how to then show the right error message based on the exception, since I think database exception is thrown for two different reasons. I agree that this is not ideal. I will think about a workaround.